js/viewmodel.js

```diff
@@ -0,0 +1,221 @@
+            pickPlacable(selectedItem);
+        }
+    }
+
+    function onItemSelected(callback) {
+        return events.on('itemSelected', callback);
+    }
+
+    // TODO: this name feels wrong
+    function pickPlacable(placeable) {
+        console.log('selected', placeable);
+        selectedItem = placeable && Object.assign({}, placeable);
+        events.emit('itemSelected', selectedItem);
+    }
+
+    function getPlaceableItems() {
+        var rawEntities = Blueprint.getEntityData();
+        return Object.keys(rawEntities)
+            .map(function (name) {
+                return Object.assign(
+                    {
+                        name: name
+                    },
+                    rawEntities[name],
+                    // fix bad data in factorio-blueprint
+                    name === 'steam_turbine' && { width: 5, height: 3 },
+                    name === 'boiler' && { width: 3, height: 2 },
+                    // fix spots where height/width are swapped
+                    name === 'pump' && { width: 2, height: 1 },
+                    name === 'decider_combinator' && { width: 2, height: 1 },
+                    name === 'arithmetic_combinator' && { width: 2, height: 1 }
+                );
+            })
+            .filter(isPlaceable)
+            .reduce(addDirectionalEntities, [])
+            .map(addDerivedProperties);
+    }
+
+    function isPlaceable(entity) {
+        return (entity.type === "item" && entity.width)
+            && entity.name !== 'curved_rail'; // curved rail is strange
+        // TODO: support tiles
+        // || entity.type === "tile"
+    }
+
+    function addDirectionalEntities(results, entity) {
+
+        if (entity.directionType) {
+            return results.concat([
+                Object.assign({}, entity, { type: 'input' }),
+                Object.assign({}, entity, { type: 'output' })
+            ]);
+        }
+        return results.concat([entity]);
+    }
+
+    function addDerivedProperties(entity) {
+        return Object.assign({}, entity, {
+            iconUrl: getIconUrl(entity),
+            canRotate: entity.width !== entity.height
+                || !entityNamesThatDoNotRotate.test(entity.name),
+            defaultDirection: getDefaultDirection(entity),
+            direction: getDefaultDirection(entity)
+        });
+    }
+
+    function getIconUrl(entity) {
+        var directionPrefix = entity.directionType
+            ? entity.type[0] + '-'
+            : '';
+
+        var fileName = directionPrefix
+            + entity.name.replace(/_/g, '-') // TODO: rename files so we can drop this
+            + '.png';
+        return 'vendor/factorio/icons/placeable/' + fileName;
+    }
+
+    function getDefaultDirection(entity) {
+        if (entity.type === 'input') { return 2; }
+        if (entity.type === 'output') { return 6; }
+        if (/inserter|offshore_pump/.test(entity.name)) { return 6; }

```

Comment:
can we move these regexes out into constants and/or arrays?

Reply:
will do